app/questionnaire/router.py

```diff
@@ -0,0 +1,52 @@
+class Router:
+
+    def __init__(self, schema, routing_path, completeness, current_location=None):
+        self._schema = schema
+        self._routing_path = routing_path
+        self._completeness = completeness
+        self._current_location = current_location
+
+    def can_access_location(self):
+        if self._is_valid_location() and not self._get_first_incomplete_block_in_section() and not self._get_first_incomplete_block_in_survey():
+            return True
+
+        return False
+
+    def get_next_location(self):
+        if not self._is_valid_location():
+            next_location = self._get_expected_location()
+            return next_location
+
+        return self._get_first_incomplete_block_in_section() or self._get_first_incomplete_block_in_survey() or self._get_expected_location()
+
+    def _is_valid_location(self):
+        return self._current_location in self._routing_path
+
+    def _get_expected_location(self):
+        next_location = self._routing_path[0]
+
+        if self._completeness.completed_blocks:
+            next_location = self._completeness.get_first_incomplete_location_in_survey()
+            if not next_location:
+                next_location = self._routing_path[-1]
+
+        return next_location
+
+    def _get_first_incomplete_block_in_section(self):
+        section = self._schema.get_section_by_block_id(self._current_location.block_id)
```

Comment:
`self._current_location` may be `None` and that'll throw an error here.

I know neither of the `_get_first_incomplete_block_in_x` methods get called if there's no `current_location` because `_is_valid_location` returns `False`. But I'd suggest being defensive here.